### PR TITLE
add entmax, for eventually improving sampling, from Sparse Text Gener…

### PR DESCRIPTION
**Pull request by lucidrains**
…ation paper

https://arxiv.org/abs/2004.02644

### CODE REVIEW

**Comment by lucidrains** [comment]
code transcribed from https://github.com/deep-spin/entmax/blob/master/entmax/root_finding.py

File: models/utils.py [hunk 1]
```diff
@@ -1,6 +1,49 @@
...
+
+        mask = mtf.greater_equal((f_m * f_lo), 0)
+        tau_lo = mtf.where(mask, tau_m, tau_lo)
+
+    if ensure_sum_one:
+        p_m = p_m / mtf.reduce_sum(p_m, reduced_dim = dim)
+
+    return p_m
+
+def sample_categorical(x, dim = None):
```
Comment (lucidrains) [inline models/utils.py:39]: can't use gumbel trick here, so had to build my own sample categorical function

can someone double check my function here?

**Comment by lucidrains** [comment]
https://github.com/deep-spin/sparse_text_generation/blob/master/dialogue_generation/eval.py#L250 If someone wants to double check the evaluation logic. I don't believe temperature needs to be used when using entmax, but I could be wrong

**Comment by sdtblck** [comment]
Looks good to me! Great work. I guess we should also add the entmax loss if possible?

**Comment by lucidrains** [comment]
> Looks good to me! Great work. I guess we should also add the entmax loss if possible?

Yup, working on it!

File: models/utils.py (hunk 1 above)
Comment (sdtblck) [reply models/utils.py:39]: this logic seems sound to me. Have you tested it out yet?

**Comment by lucidrains** [comment]
ents

File: models/utils.py (hunk 1 above)
Comment (lucidrains) [reply models/utils.py:84]: yup, i do so at the end of test_models!

File: models/utils.py [hunk 2]
```diff
@@ -1,5 +1,93 @@
...
+    max_val = mtf.reduce_max(x, reduced_dim = dim)
+
+    tau_lo = max_val - _gp(1, alpha)
+    tau_hi = max_val - _gp(1 / d, alpha)
+
+    f_lo = mtf.reduce_sum(_p(x - tau_lo, alpha), reduced_dim = dim) - 1
+
+    dm = tau_hi - tau_lo
+
+    for _ in range(n_iter):
```
Comment (lucidrains) [inline models/utils.py:39]: am i supposed to be using `mtf.while_loop` here?

Comment (sdtblck) [reply models/utils.py:39]: I think that a for loop is fine here? 
Using mtf.assign() might be more efficient than dm = dm/2, since a variable will get added to the graph for each iteration in the for loop here. Probably not a problem if n_iters stays around 50 though.

**Comment by sdtblck** [comment]
is this ready to merge @lucidrains ?

**Comment by lucidrains** [comment]
@sdtblck yea i think so!

**Comment by sdtblck** [comment]
Done! Thanks as always
